### DISCUSSION TITLE
Consistant video file formatting?

### DISCUSSION THREAD

**Discussion by gateway** [rkhamilton/vqgan-clip-generator#15] [General]
It seems that we might have an issue with the filename changes for video export when running this though the restyling option.

Current exported file names from the video are frame_%12d.jpg which is awesome esp for any big video file extraction.

When you run this though restyle_video_frames it converts the file name into 1.png, 2.png 3.png etc.. and then if you use upscaling it outputs the same 1.png , 2.png etc

However when running the ffmpeg commands its looking for filenames that are in this format frame_%12d.jpg

I manually used ffmpeg to create my video which was fine.  I'm not sure if i missed a setting in video restyle to keep proper filenames?

here is some hacky code I was messing around while working on other stuff.

```from vqgan_clip import generate, video_tools, esrgan
from vqgan_clip.engine import VQGAN_CLIP_Config
import os

config = VQGAN_CLIP_Config()
config.output_image_size = [587,330]
config.init_weight = 1.0
text_prompts = 'portrait on deviantart'
input_video_path = 'rory.mp4'
final_output_filename = os.path.join('output','output.mp4')
copy_audio = True
extraction_framerate = 30
extracted_video_frames_path = 'video_frames'
output_framerate = 30
upscale_image = True

# Use a wrapper for FFMPEG to extract stills from the original video.
original_video_frames = video_tools.extract_video_frames(input_video_path, 
        extraction_framerate = extraction_framerate)

# Apply a style to the extracted video frames.
generate.restyle_video_frames(original_video_frames,
        eng_config=config,
        text_prompts = text_prompts,
        iterations = 20,
        save_every=None,
        current_source_frame_prompt_weight=0.1,
        previous_generated_frame_prompt_weight=0.0,
        generated_frame_init_blend=0.2)

#Upscale image 
if upscale_image:
    esrgan.inference_realesrgan(input='.\\video_frames',
        output_images_path='upscaled_video_frames',
        face_enhance=False,
        purge_existing_files=True,
        model_filename='RealESRGAN_x4plus_anime_6B.pth',
        model_url='https://github.com/xinntao/Real-ESRGAN/releases/download/v0.2.2.4/RealESRGAN_x4plus_anime_6B.pth')


# Encode the video.
generated_video_no_audio=os.path.join('output','output_no_audio.mp4')
os.system(f'ffmpeg -y -f image2 -i upscaled_video_frames\frame_%12d.jpg -r 30 -vcodec libx264 -crf 23 -pix_fmt yuv420p -strict -2 output_no_audio.mp4')

# Copy audio from the original file
if copy_audio:
        video_tools.copy_video_audio(input_video_path, generated_video_no_audio, final_output_filename)
        os.remove(generated_video_no_audio)
else:
        os.rename(generated_video_no_audio,final_output_filename)```

**Comment by rkhamilton** [comment]
I agree with you that this is not the best behavior. Restyle_video_frames should preserve the original filenames and use them in the new output directory. I started to just make that change, but a number of other things are impacted that need to be untangled. It's not feeling like I'm saving the user any complexity at this point, as well.

With version 1.1 I've split the encoding out to separate lines in the user script, rather than burying it inside of the main restyle_video method. I'm feeling like the encode_video and extract_video methods are more trouble than they are worth at this point, and may have never been a good idea. They are adding a lot of code fragility to avoid having a single line call of os.system() in the user's script. I'm leaning toward just removing the video tools methods completely and putting example os.system calls to ffmpeg in the examples files instead. Then it's very transparent to users what is happening, and they can freely change ffmpeg parameters, encoders, etc, without me having to explicitly code a bunch of flexibility into it.

You can see an example of such an example script right now in [Real-ESRGAN.md](https://github.com/rkhamilton/vqgan-clip-generator/blob/main/Real-ESRGAN.md). My wrappers aren't even saving any lines of code for the user. What would you think about just dropping the wrappers and expecting people to use os.system? My examples would include f-strings so you could see how to easily include variables from your script. It would look something like the example below.
```python
os.system(f'ffmpeg -y -f image2 -i {input_frames_path} -r 30 -vcodec libx264 -crf 23 -pix_fmt yuv420p -strict -2 {output_file}')
```

**Reply by gateway** [reply]
While I agree with most of what your saying and your maintaining the code base with updates, fixes and new features so I can relate.  However you have done some of the hard work already and it would seem like the proper file name frame_%12d.jpg or frame_%12d.png shouldn't be a big change across the other functions? 

Yes leaving it to os calls is always easier but a lot of people have to spend time figuring out how to use ffmpeg properly :) lol.

So far the only thing I saw was that restyle video which I'm sure goes across zoom and the other video functions just needs to not rename the files and maybe adding a config switch for config.esrgan = true (meaning to look in the upscale folder to process the videos) ?

Honestly im not sure how much work it would be but for a lot of people maybe just wanting to play this saves them a ton of figuring stuff out.. anyhow just thought I would throw it out and let you know what I ran into so far.

**Reply by gateway** [reply]
sorry I forgot to mention it took me a minute to figure out how to handle files that are named like 1.png , 2.png and encoded them.. I had to look at your code since Im used to file_%12.png.. didn't realize I could use %d.png :) 

I should mentioned I'm only digging in slowly when I have a sec to fire up my ai box.. will write some of my own code this weekend to use and test your stuff.

**Reply by rkhamilton** [reply]
I've pushed v1.1.3 which has the changes to restyle_video* so that it preserves original filenames. Extracting frames now uses the format frame_%12d.png, and the video_tools encode method requires that format.

Also, since I was frequently interested in tweaking my ffmpeg run, now whenever you encode a video with encode_video, at the end it will print the exact command string that was used so you can copy/edit/paste that into another command prompt if desired.

**Reply by gateway** [reply]
Ill give it a shot.. I just finished up a video from your settings with 1.1.2 and its hard for me to keep the shape/look of my friends face.. probably to much going on.. also have you tested deflicker options in ffmpeg before? http://underpop.online.fr/f/ffmpeg/help/deflicker.htm.gz

thanks for changing the filenames!!!

**Comment by gateway** [comment]
I think I have the latest code @rkhamilton but when doing tests with generate.multiple_images function its outputting images at 1.png , 2...
just a fyi

**Reply by rkhamilton** [reply]
Thanks for letting me know. I didn't think of multiple_images as generating content for video, but rather for generating multiple standalone images you can pick your favorite RNG output. But it's inconsistent behavior, so I changed it to frame_%12d.png for the next version.

**Reply by gateway** [reply]
This relates to a new question i had about that function.. lol.. https://github.com/rkhamilton/vqgan-clip-generator/discussions/22
